### ISSUE TITLE
Admin bar menu item may not appear in some cases even when data exists

### ISSUE THREAD

**Issue by aaemnnosttv** [google/site-kit-wp#2508]
## Bug Description

The Analytics and Search Console module currently both implement a protected `has_data_for_url( $url ) : boolean` method which is only used right now to help determine if the Site Kit admin bar menu item should be added at page load.

This method queries the respective module's API to check for some minimal data and caches the result in a transient for the given `$url` by hashing the URL with `md5`.

Because the transient cache key is only based on the URL, these methods have the potential to result in an inaccurate result if for example, the property changed once the result had been cached.

A successful result (has data) is cached for 1 day, while a no-data result is only cached for 1 hour.

To avoid this potential inaccurate behavior, we should update the cache key to also include other values which are dependencies of the request which is made, particularly any settings-based values which can change which are referenced in `create_analytics_site_data_request` and `create_search_analytics_data_request` respectively.

For Analytics, this would be the profile ID and for Search Console this would be the property ID.

This is more relevant for Analytics since the Search Console property can't change once the plugin is set up under normal circumstances, but it would help avoid false positives when testing the plugin in non-standard ways as well as be a generally more accurate way to cache the results.

Alternatively, we could remove such server-side checks and make these requests from the client side instead of conditionally adding our admin bar menu item.

## Steps to reproduce

1. Identify an Analytics account you have access to that has data
1. Set up Site Kit using a domain / search console property that **does not have data**
1. Activate Analytics using a new account or one that has **no data**
1. Navigate to a post that has data for that URL in the account identified in step 1
    - Note the Site Kit admin bar menu item is not present on the front end for this post
1. Change the Analytics settings to now use the account **with data**
1. Navigate to the same post that has data for that URL in the account identified in step 1
    - Note the Site Kit admin bar menu item is still not present on the front end for this post

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* The Site Kit admin bar menu should cater for the case that only responses with "zero data" state are returned. If so, the numbers and More Details link on the admin bar menu should be replaced with the following content:
    * Title: _No data available yet_
    * Description: _There is no data available for this content yet. This could be because it was recently created or because nobody has accessed it so far._
* There is no exact design for this one, so let's just use fonts and sizes like we're using elsewhere and review the details during CR and testing. It should _not_ use the blue boxes that we have in the dashboard though, it should just be the regular black text color in front of the white background of the admin bar menu.
* The PHP logic to check whether a post has Search Console or Analytics data should be entirely removed.
* From the PHP side, the admin bar menu should be loaded/displayed for every public post, regardless of whether it has data or not.
* A Storybook story should be added for when the admin bar menu shows "zero data" state.

## Implementation Brief

**JS Changes**

**`AdminBarDataZero` Component**

* Create a new component in the `assets/js/components/adminbar` directory called `AdminBarDataZero`.
   * This component should render the following text with the same structure and layout as a normal admin bar widget with data (eg. assets/js/modules/analytics/components/adminbar/AnalyticsAdminbarWidgetOverview.js)
             * Title: _No data available yet_
             * Description: _There is no data available for this content yet. This could be because it was recently created or because nobody has accessed it so far._

**Detect No Data State to show `AdminBarDataZero`** 

 * Based on the "Error & No Data States" section of the IB for [1921](https://github.com/google/site-kit-wp/issues/1921) we should expose a `hasData` function on `AdminBarImpressions`, `AdminBarClicks`, `AdminBarUniqueVisitors` and `AdminBarSessions`. This function should return a select function with the logic to select the data for that component based on the current store state and return a boolean from the result of this components data passed to it's `isZeroReport` function. This prevents us from duplicating selectors in the `AdminBarApp` and it's children.
 * Once the `hasData` function has been implemented on each of the child components we can check if they have data using:

```
const adminBarImpressionsHasData = useSelect( AdminBarImpressions.hasData );
const adminBarClicksHasData = useSelect( AdminBarClicks.hasData );
const adminBarUniqueVisitorsHasData = useSelect( AdminBarUniqueVisitors.hasData );
const adminBarSessionsHasData = useSelect( AdminBarSessions.hasData );
```

 * If **none of the child components have data** (`const dataZero = (! adminBarImpressionsHasData && ! adminBarClicksHasData && ! adminBarUniqueVisitorsHasData && ! adminBarSessionsHasData);`), render the new `AdminBarDataZero` in place of all of the components inside the [`<div className="mdc-layout-grid__inner">` div L81-L107](https://github.com/google/site-kit-wp/blob/5fdf25f860ed77d24994562af84d94b2333ebaa3/assets/js/components/adminbar/AdminBarApp.js#L81-L107).

**Storybook**

 * Create a new storybook story under the [Admin Bar story](http://localhost:9001/?path=/story/global--admin-bar) called `Admin Bar Data Zero` that mocks up and shows the admin bar `AdminBarDataZero` component in this new state.

**PHP Changes**

* Update `includes/Modules/Analytics.php`
   * Remove `is_active_in_admin_bar()` function entirely.
   * Remove the `has_data_for_url()` function entirely.
* Update `includes/Modules/Search_Console.php`
   * Remove `is_active_in_admin_bar()` function entirely.
   * Remove the `has_data_for_url()` function entirely.
* Update `includes/Core/Admin_Bar/Admin_Bar.php`
   * Replace the outdated `Module_With_Admin_Bar` check from `is_active()`:
   * https://github.com/google/site-kit-wp/blob/a0976b218f76d3e705dee7d6091f8945629862aa/includes/Core/Admin_Bar/Admin_Bar.php#L205-L225
   * Keeping the filter intact:
```
   /**
    * Filters whether the Site Kit admin bar menu should be displayed.
    *
    * The admin bar menu is only shown when there is data for the current URL and the current
    * user has the correct capability to view the data. Modules use this filter to indicate the
    * presence of valid data.
    *
    * @since 1.0.0
    *
    * @param bool   $display     Whether to display the admin bar menu.
    * @param string $current_url The URL of the current request.
    */
    return apply_filters( 'googlesitekit_show_admin_bar_menu',true, $current_url );
```
   * Remove the class use statement: `use Google\Site_Kit\Core\Modules\Module_With_Admin_Bar;`
* Remove the `use` and `implements` reference to `Module_With_Admin_Bar` from both:
   * includes/Modules/Analytics.php
   * includes/Modules/Search_Console.php
* Remove the now unused class file `includes/Core/Modules/Module_With_Admin_Bar.php`.

* Delete the `tests/e2e/plugins/admin-bar-visibility.php` file as we no longer need to override this for testing as it will now be displayed on all posts by default.
* Remove `await activatePlugin( 'e2e-tests-admin-bar-visibility' );` from all of these files:
    * tests/e2e/specs/amp/admin-bar.test.js
    * tests/e2e/specs/amp/home-validation.test.js
    * tests/e2e/specs/modules/analytics/admin-bar.test.js
    * tests/e2e/specs/modules/search-console/admin-bar.test.js

### Test Coverage

* Ensure the e2e tests that reference the admin bar still pass:
    * tests/e2e/specs/amp/admin-bar.test.js
    * tests/e2e/specs/amp/home-validation.test.js
    * tests/e2e/specs/modules/analytics/admin-bar.test.js
    * tests/e2e/specs/modules/search-console/admin-bar.test.js

### Visual Regression Changes

* No VRT changes.

## QA Brief

* Reset your site kit data and re-connect making sure to set your domain in the Developer Settings plugin to https://www.elasticpress.io/
* Create a page in your local site called /features:
   * Confirm the admin bar on this page shows the correct admin bar with data:
   * <img width="1017" alt="Screenshot 2021-01-13 at 11 43 38" src="https://user-images.githubusercontent.com/119352/104448081-9f3c9300-5594-11eb-96b7-f1722ccbf24e.png">
   * Set up the search console and analytics modules and confirm the notices show correctly in each case.
* Create a page *which doesn't exist on the elasticpress.io site* (eg. /test-page-no-data/) in your local site:
   * Confirm that the admin bar now shows on this page (where previously it would be hidden) and that it shows the new Zero Data component:
   * <img width="1018" alt="Screenshot 2021-01-13 at 11 43 47" src="https://user-images.githubusercontent.com/119352/104448293-df9c1100-5594-11eb-9318-036a64fe5b0c.png">
   * Confirm the design and copy match the AC.

## Changelog entry

* Always display the Site Kit admin bar menu even when there is no data for a more consistent experience, informing about the lack of data as applicable.

**Comment by adamsilverstein** [comment]
> Alternatively, we could remove such server-side checks and make these requests from the client side instead of conditionally adding our admin bar menu item.

This approach would also resolve https://github.com/google/site-kit-wp/issues/2528

**Comment by aaemnnosttv** [comment]
@felixarntz – I'm wondering if we should do these server-side checks at all and defer it to the client instead. We don't hide other components entirely when there is no data yet, so it's a bit odd that we would do that with the admin bar app. Especially if this has the potential to result in slow page loads, that would be quite poor to impose on every front end page load. The app is already lazy loaded on click/hover so there would be better from a quota perspective as well.

**Comment by archon810** [comment]
You can do them in cron, or in another async manner, like Ajax, but please not in a synchronous way that massively delays wp admin responses.

**Comment by felixarntz** [comment]
@marrrmarrr I've added preliminary wording in the ACs here - can you revise please?

**Comment by felixarntz** [comment]
@aaemnnosttv Changed this to an enhancement as that's more applicable to the change we're actually making here (even though it also fixes the bug you've mentioned).

**Comment by benbowler** [comment]
Added an IB for approval.

**Comment by aaemnnosttv** [comment]
@benbowler IB looks good for the most part, just a few points to address:

> Create a new component in the `assets/js/components/adminbar` directory called `ZeroDataWidget`.

This isn't really a widget; just a component like `ReportZero` is. Let's call it `AdminBarDataZero` for consistency but also prefixed for consistency with the `adminbar` components since it isn't a generic zero data component to be used anywhere.

> Update `assets/js/modules/analytics/components/adminbar/AnalyticsAdminbarWidgetOverview.js`
> Update `assets/js/modules/search-console/components/adminbar/SearchConsoleAdminbarWidgetOverview.js`

Note that some admin bar components are changing in #2543 (marked as a blocking dependency of this issue and is soon to be merged) and there are now new versions of the components that render the data which did not exist at the time this issue was originally defined. It may complicate the zero data detection though if we need to account for both legacy and modern components.

More importantly, I think there is a misunderstanding here in that the `AdminBarDataZero` component wouldn't be used in the individual data/widget components, but instead rendered in place of the widgets and "more details" link when ALL of the data is zero data. That is, this component would only show in the case where all of the others are in a zero-data state to prevent the all-blue-boxes grid. This is similar to what we do on the dashboard details screen when it is loaded for an entity that isn't found on the current site. 
![image](https://user-images.githubusercontent.com/1621608/103630938-0c27ab80-4f4b-11eb-95d2-eca8f1503c19.png)

> Remove `is_active_in_admin_bar()` function entirely.

This isn't possible without also removing the `Module_With_Admin_Bar` interface from the class. I'm not sure if this makes sense to keep any more or not since it was mostly added as an abstraction to decouple the Analytics and SC module classes from `Admin_Bar`. It's no longer needed for our current use since the admin bar item will no longer conditionally be added based on module data. There may be other reasons to keep it, but I'm thinking this can probably be removed as well. @felixarntz 

> Delete the `tests/e2e/plugins/admin-bar-visibility.php` file as we no longer need to override this for testing and we have removed the `googlesitekit_show_admin_bar_menu` filter.

We would also need to remove instances where this is activated in E2E tests (i.e. `activatePlugin( 'e2e-tests-admin-bar-visibility' )`).

---

There's a few points where we need input from @felixarntz but the rest can be iterated on while we're waiting for him to clarify those details 👍

**Comment by benbowler** [comment]
Thanks @aaemnnosttv, I've updated it IB based on your suggestions. I'll assign to @felixarntz for his thoughts on removing `Module_With_Admin_Bar`. I can't see it used anywhere else in the code right now.

**Comment by felixarntz** [comment]
@aaemnnosttv @benbowler Good point, let's remove the `Module_With_Admin_Bar` interface and related usage.

The `googlesitekit_show_admin_bar_menu` filter should be kept though. I think that's what you were thinking so far anyway, but just wanted to clarify. While we won't be using it anymore, it should be there for 3P developers in case they don't want the admin bar menu entirely for example.

**Comment by aaemnnosttv** [comment]
Thanks @felixarntz 👍 

@benbowler let's remove the `Module_With_Admin_Bar` interface entirely (only used by Analytics and SC modules) along with the usage as you've already described. As Felix said, we need to keep the `googlesitekit_show_admin_bar_menu` filter on the return value of the `Admin_Bar::is_active` method.

**Comment by benbowler** [comment]
Hey @aaemnnosttv, I've updated the IB with the hasData function on each child component as per our call.

**Comment by aaemnnosttv** [comment]
@benbowler IB looks good to me, except the Test Coverage and VRT sections are still empty. Would you please fill these in even if there are no changes anticipated?

**Comment by benbowler** [comment]
Updated to add Test Coverage and VRT notes. @aaemnnosttv

**Comment by aaemnnosttv** [comment]
@benbowler IB ✅ 

> The visual tests should be updated to include the new Admin Bar Data Zero story.

We just need the added story, no added VRT test here just yet. Not sure if that's what you meant but otherwise good to go!

**Comment by benbowler** [comment]
I've implemented the core of this ticket and I'm just waiting on #2594 before I implement the required storybook story.

**Comment by benbowler** [comment]
QA Brief added.

**Comment by benbowler** [comment]
Story added and ready for Code Review.

**Comment by benbowler** [comment]
Thanks for your review @aaemnnosttv, I've made significant changes to the selector and simplifications to the layout.

The main admin bar story is now showing the No Data component too, I'm looking at that now, I didn't want to block further review with that.

**Comment by aaemnnosttv** [comment]
@felixarntz – I missed a few details with the IB which was too-specific to the new widget components and didn't account for legacy widgets. Given the timing around the release for this one, I've created a PR on top of @benbowler 's which addresses this. Assigning to you for review.

**Comment by felixarntz** [comment]
Approved the extra PR, but I saw a couple things that might require broader changes, and already the existing PR makes some complex changes, so I'll move this to the next release.

**Comment by benbowler** [comment]
Hey @aaemnnosttv, can you evaluate my approach over on the [PR #2607](https://github.com/google/site-kit-wp/pull/2607). I've worked to Felix's points.

I had an issue with a render loop because `setWidgetState` is called on a `useEffect` in `useWidgetStateSideEffect`. This lead to the following:

1. `AdminBarWidgets` renders each child widget.
1. These widgets all have no data, loading the `WidgetZeroData` component which triggers setWidgetState for each.
1. `AdminBarWidgets` detects the no data state on all widgets and renders `AdminBarZeroData`.
1. This intern de-renders all of the child widgets, resetting the getWidgetState and the `AdminBarWidgets` no longer detects no data state starting the cycle again.

I solved this using the `HIDDEN_CLASS` to hide sections that have no data without de-rendering them.

Let me know your thoughts.

**Comment by benbowler** [comment]
In code review based on discussion with @aaemnnosttv.

**Comment by felixarntz** [comment]
@benbowler 
> I solved this using the `HIDDEN_CLASS` to hide sections that have no data without de-rendering them.

Yes, this is what we're doing for the "real" widgets itself. It feels a bit weird, but for now it's a good enough solution to achieve our goal which is to have the special widget states (e.g. no data) being fully driven by the rendering component, without requiring calls to specific datastore actions manually. This is mostly so that 3P developers can simply write their components and don't need to worry about specific datastores. Our documentation will explain that they should return the specific `Widget*` components for special state.

**Comment by aaemnnosttv** [comment]
@benbowler – assigning back to you to resolve the conflicts with the branch.

**Comment by aaemnnosttv** [comment]
CR ✅  from me; Assigning to @felixarntz since he previously requested changes.

**Comment by cole10up** [comment]
### QA ✅ 

**Tested**
Confirmed functionality of the new admin bar on all pages on my personal site:
![image](https://user-images.githubusercontent.com/57115907/107607862-442ba980-6bef-11eb-8174-b6fc61f2812d.png)

![image](https://user-images.githubusercontent.com/57115907/107607921-6291a500-6bef-11eb-9006-2380ec015afa.png)

![image](https://user-images.githubusercontent.com/57115907/107607941-6de4d080-6bef-11eb-8eec-871adf10d50a.png)

Created a new test page:
![image](https://user-images.githubusercontent.com/57115907/107607973-80f7a080-6bef-11eb-8414-6e352c5a0cfb.png)

Looks good on my end, sending to Approval
